docs: Add Committer Expectations document

keith-zephyr opened the pull request:
Collect up all the committer expectations and PR requirements into a single place. Add additional guidelines about creating small PRs and how to break up PRs into multiple commits.

Signed-off-by: Keith Short <keithshort@google.com>

keith-zephyr committed:
docs: Add Contributor Expectations document

Collect up all the contributor expectations and PR requirements into a
single place. Add additional guidelines about creating small PRs and how
to break up PRs into multiple commits.

Signed-off-by: Keith Short <keithshort@google.com>

keith-zephyr commented:
Marked as a draft, but this is ready for feedback. Once the content has of this doc has consensus, I'll also update the guidlines.rst to remove redundant/obsolete info and link to this doc.

keith-zephyr commented:
https://github.com/zephyrproject-rtos/zephyr/issues/50836

MaureenHelm commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,165 @@
+  feature. In this case create an :ref:`RFC proposal <rfcs>` [#rfc]_ to describe
+  the additional parts of the feature for reviewers.
+
+- PRs should include tests for any added or changed lines of code.
```
```suggestion
- PRs should include tests or samples for any added or changed lines of code.
```

teburd replied:
Is there a particular goal you have in mind by trying to require testing here? Many contributions to Zephyr come from one off small driver additions/features/bugfixes from contributors working on a product. There's already the remarkably high barrier to entry where many PRs die today of simply passing git/code linting steps. Adding yet another brick on the wall feels like we'll lose those first time/small contributors even faster. Sensors are especially prone to this sort of thing. There's over a 100 sensor drivers that more or less are community contributed and tested, with the project providing samples to build against and run.

Speaking of which if a PR contains a bug fix for a piece of hardware what is the real expectation here? If we look at i2c drivers for example very few are actually being tested on hardware given the i2c test case requirement of having a specific external IC connected to it.

Testing by building the driver in a sample or test is about the same, its a smoke test to make sure at least the code builds which is 90% of what we get with CI today when you exclude simulators.

If you consider that, samples and tests are about equivalent in that regard.

keith-zephyr replied:
IMO, the Zephyr project should move towards 100% test coverage, including drivers.  I think the expectations for tests should be:
1. When fixing a hardware-agnostic bug, PR authors should create a new test that passes with the fix applied, but fails without the fix.
2. When fixing a hardware specific bug (SoC driver, peripheral driver, etc), the PR author should fix the corresponding tests/emulators as well.  If the driver doesn't currently have any tests or emulation, this requirement is waived.  I agree it isn't reasonable to hold up a bug fix and require the PR author to write a new emulator when one doesn't exist.
3. When introducing a new driver, one should also contribute an emulator and tests for the driver.  The PR author has already done a deep dive into the datasheet and has the expertise to write an emulator at the same time.  I fully expect that there will be latent bugs introduced because the PR author writes both the driver and emulator.  However, once the driver is run on real hardware any driver bug fixes will require a fix to the emulator (see previous comment).

MaureenHelm commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,165 @@
+
+- Small PRs should encompass one self-contained logical change.
+
+- When adding a new feature or API, the PR should address only one part of the
```
```suggestion
- When adding a new large feature or API, the PR should address only one part of the
```

Some new features may be small and not worth splitting up.

ElectricalPaul commented on doc/contribute/committer_expectations.rst:
This seems like a particularly good idea, instead of losing history when someone amends the commit and force-pushes.

mbolivar-nordic replied:
I'm familiar with `git rebase -i --autosquash` and I use it frequently on my local repositories, but I think this may be unnecessary given the previous bullet. In my experience, separate force-pushes for rebases and code changes show up cleanly enough in the github diffs to make review tolerable.

alevkoy replied:
In case anyone missed it before, see https://github.com/zephyrproject-rtos/zephyr-testing/pull/203. Right now, I have a couple fixup commits to respond to review comments, and I'm waiting for people to review those before I squash.

mbolivar-nordic commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,165 @@
+
+- Reviewed more quickly and reviewed more thoroughly. It's easier for reviewers
+  to set aside a few minutes to review small changes several times than it is to
+  allocate a 30 minute block to review a large PR.
```
It seems minor, but I think it is worth discussing whether this number should remain here.I don't want people to get the mistaken impression that 30 minutes is the worst case scenario for a reviewer. 

In my experience, some reviews take multiple hours or even days if you're going to do them carefully and not just rubber-stamp.

I would suggest something like "allocate a large block of time" for parallelism with the "review a large PR" language.

mbolivar-nordic commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,165 @@
+  to set aside a few minutes to review small changes several times than it is to
+  allocate a 30 minute block to review a large PR.
+
+- Less wasted work if the PR if reviewers or maintainers reject the direction of
```
'if the PR if'?

mbolivar-nordic commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,165 @@
+- Small PRs should encompass one self-contained logical change.
+
+- When adding a new feature or API, the PR should address only one part of the
+  feature. In this case create an :ref:`RFC proposal <rfcs>` [#rfc]_ to describe
```
What is the `[#rfc]_` part doing here?

mbolivar-nordic replied:
never mind, resolving this one

mbolivar-nordic commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,165 @@
+Multiple Commits on a Single PR
+===============================
+
+You are further encouraged to break up PR into multiple commits.  Keep in mind
```
PRs

mbolivar-nordic commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,165 @@
+- Submitting a new feature.
+- Submitting a new API.
+- :ref:`treewide-changes`.
+- Other changes which encompass more than 200 lines of non-trivial modifications.
```
Curious how you arrived at this number as well

mbolivar-nordic commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,165 @@
+- When major new functionality is added, tests for the new functionality MUST be
+  added to the automated test suite. All new APIs MUST be documented and tested
+  and tests MUST cover at least 80% of the added functionality using the code
+  coverage tool and reporting provided by the project..
```
extra period at end of sentence

mbolivar-nordic commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,165 @@
+
+- When major new functionality is added, tests for the new functionality MUST be
+  added to the automated test suite. All new APIs MUST be documented and tested
+  and tests MUST cover at least 80% of the added functionality using the code
```
Do we have some way to measure this? We used to have codecov.io, but that ended up a dud

mbolivar-nordic commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,165 @@
+  coverage tool and reporting provided by the project..
+
+- Incompatible changes to APIs must also update the release notes for the
+  next release detailing the change.  APIs marked as experimented are excluded
```
experimental

mbolivar-nordic commented on doc/contribute/committer_expectations.rst:
I continue to want to caution you that there are several areas in the project where review bandwidth is extremely underprovisioned. Big examples are the build system and devicetree bindings. We all want prompt review, but we need to also be realistic about the resources we have as an open source project. I continue to be in favor of prompt review for roadmap items approved by the TSC, but there just isn't enough time in all cases to try to do this project-wide IMO.

alevkoy replied:
+1. I think a well understood escalation path is probably more valuable than a very fast review SLO.

keith-zephyr replied:
Leaving open for discussion by the Process working group.

mbolivar-nordic commented on doc/contribute/committer_expectations.rst:
Sometimes I see people splitting up reviews in two:

1. first review: nitpicks, code style change requests, etc
2. second: substantive technical comments

This can be useful especially when the nitpick is something like "this PR is doing X inconsistently throughout all its changes and it's making it harder to review; please fix it and I'll take a look at the heart of the changes".

Thoughts on things like that?

alevkoy replied:
@mbolivar-nordic: I think the two-stage review you describe is a reasonable review strategy. I think it's very appropriate to send an initial comment to the effect of, "You need to completely restructure this," and then not look at it again until they do to avoid getting into style issues or naming things in code that will go away anyway.

@keith-zephyr: I think reviewing the whole PR is a good way to go from two round trips worth of review to one round trip. For PRs where the reviewer is asking for high-level changes, it seems unlikely that there will just be one round trip, and the reviewer's and author's time might be better spent by getting it right at a high level on one pass before worrying about details. I'm inclined to say this is a judgment call on the part of the reviewer.

alevkoy replied:
I don't know what the best way to express this in a document is, but the cost of a partial review in increased round trips goes down when the reviewer is more responsive. To a certain extent, reviewing an entire PR probably increases the round trip time compared to reviewing part of it. I'd rather work with a reviewer who responds one commit or file at a time, every few days, in a large PR than one who avoids the review for weeks at a time, loathing and dreading how long it's going to take them to review the whole thing.

keith-zephyr replied:
Updating language to call out good examples of partial reviews.

mbolivar-nordic commented on doc/contribute/committer_expectations.rst:
I think this is an excellent mechanism for avoiding drive-by feature creep requests

fabiobaltieri replied:
The opposite may be good: ask to decrease the scope of the PR if it makes sense, maybe split some commit on a separate PR.

alevkoy replied:
+1

keith-zephyr replied:
Regarding decreasing scope, that is mentioned above just below the PR requirements.  Should I move that into the reviewer expectations.

mbolivar-nordic commented on doc/contribute/committer_expectations.rst:
And enter into a fight that you can't win against zephyr-bot? I think this is impractical.

keith-zephyr replied:
Zephyrbot assigns too many reviewers in my opinion.  I was looking for a mechanism for the reviewers to signal to the author (and other reviewers) that they do not plan on looking at a PR.  Removing this language for now.

Only adding a comment stating you don't plan to review doesn't really help move a PR forward.  Github will also continue to spam the uninterested reviewer on any comments or force pushes.

mbolivar-nordic reviewed (commented):
Overall seems reasonable. I know this is draft, but I'll leave some nitpicks alongside substantive points just in case.

Are you planning on reorganizing other documents too to make them easier to find? Or when you take this out of draft, will it still be mainly building on top of existing docs?

Thanks for this!

alevkoy commented on doc/contribute/committer_expectations.rst:
@keith-zephyr It might be good to encourage reviewers to call out comments that aren't high-priority for the reviewer and to approve PRs with open comments when they consider the remaining disagreements trivial. (If the author agrees, they can fix it, and the reviewer will presumably approve it again.) That would eliminate a lot of back and forth, and it would decrease the sense of doom on the part of the author when reviewers keep finding new things to comment about.

alevkoy replied:
I think so. I would suggest calling that out here and noting that you should "approve" instead of "request changes" if all your remaining comments are non-blocking. Personally, I was surprised by the implications of "request changes" as a new Github user.

teburd reviewed (commented):
I generally love that we are trying to write this down. There needs to some compromise though I think given many contributions and contributors come from using zephyr to finding a bug and submitting a small update. These contributors should be encouraged with easy to submit and contribute changes with as few barriers to entry as possible.

marc-hb commented on doc/contribute/committer_expectations.rst:
Not sure what is the best place but I have explained to very many people new to Github that they too can use the "Start Review" button on their _own_ Pull Requests to reduce notification spam by one order of magnitude - and also get a chance to proof-read their comments.

marc-hb replied:
Added and commented above, resolving here.

mbolivar-nordic commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,248 @@
+Large Changes
+=============
+
+Large changes to the Zephyr project must submit an :ref:`RFC proposal <rfcs>`
+describing the full scope of change and future work.  The :ref:`RFC proposal
+<rfcs>` provides the required context to reviewers, but allows for smaller,
```
Minor: is there a reason for doing two `:ref:`s to the same link in the same paragraph? Seemed a bit odd to me

mbolivar-nordic commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,248 @@
+      kernel timer.
+    - TODO: Provide a driver level API test example.
+
+- Incompatible changes to APIs must also update the release notes for the
+  next release detailing the change.  APIs marked as experimental are excluded
```
This will be worth highlighting to maintainers; we don't do this now.

mbolivar-nordic commented on doc/contribute/committer_expectations.rst:
typo: reviewers

marc-hb commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,248 @@
+
+- Follow the Zephyr :ref:`coding_style` and :ref:`coding_guidelines`.
+
+- PRs must pass all CI checks. This is a requirement to merge the PR, but you
+  may mark a PR as draft and explicitly request reviewers to provide early
```
> first-time contributors require an org member to approve the CI run for their PRs.

While a little documentation never hurts, does the "user-interface" make this and the next step(s) clear? It really should.

keith-zephyr commented:
Commit https://github.com/zephyrproject-rtos/zephyr/commit/16bdac253c34019a2dbcf66a89bacff8231a2085 is a rebase only.

alevkoy commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,272 @@
+  number of emails sent to reviewers.
+
+- Try to minimize rebases in the middle of a review. If a rebase is required,
+  push this as a separate update with no functional changes to the PR. When you
```
Non-blocking: I would say, "no other changes since the last push of the PR." Many changes of the type that people ask for in code reviews will not be functional.

fabiobaltieri commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,183 @@
+Overview
+********
+
+The Zephyr project encourages you to submit changes as small pull requests.
```
"Small" may not be the perfect word here but I reckon it's the easiest to pass the idea to an inexperienced user, people that don't do many code reviews have no clear idea of what "easily reviewable" means, and while it's explained right below, I think that the idea of "small" is a good enough starting point for readers that don't get to the details explanation and can't wait to hit the "Open PR" button. But to your point, the kernel documentation does not quite mention "small" in the context of a commit, though it sort-of does in term of a patchseries: https://docs.kernel.org/process/submitting-patches.html#split-changes

marc-hb replied:
> The "Trivial" label is used to mark PRs that are not subject to the 2 day review period.

Found it in doc/project/dev_env_and_tools.rst , thanks.

marc-hb replied:
I just realized most of the text that follows is about "smaller" PRs rather than "small" PRs that we've been struggling to define. This switch would IMHO resolve a few discussions below. In fact the word "smaller" is already used a few times.

```suggestion
The Zephyr project encourages you to submit changes as smaller pull requests.
```

mbolivar-nordic commented:
@fabiobaltieri ping

marc-hb commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,183 @@
+Defining Small PRs
+==================
+
+- Small PRs should encompass one self-contained logical change.
```
"One self-contained logical change" is already a very common terminology for commits[*]. So this line is practically requesting single-commit PRs. Is that intended?

[*] https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes

marc-hb commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,165 @@
+========================================
+
+- Don’t resolve comments until the question has been answered or the request
+  completed in an update to the PR.
```
"Unless they applied the reviewer's recommendation exactly, authors must not resolve and hide comments, they must let the initial reviewer do it. The Zephyr project does not require all comments to be resolved. Leaving some completed discussions open can sometimes be useful to understand the greater picture".

marc-hb commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,272 @@
+- Don't resolve comments until the question has been answered or the request
+  completed in an update to the PR.
+
+- Respond to comments from the "Files changed" view. This allows you to respond
```
```suggestion
- Respond to comments using the "Start Review" and "Add Review" green buttons in the "Files changed" view. This allows you to respond
```

It's possible to spam one notification per comment (a github bug TBH...) from the "Files Changed" tab.

marc-hb commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,272 @@
+  more collaboration as it is easier for multiple contributors to work on a
+  feature once the minimum implementation has merged.
+
+- Mark trivial, non-functional, requests as "Non-blocking" in your comment.
```
```suggestion
- When using the "Request Changes" option, mark trivial, non-functional, requests as "Non-blocking" in your comment.
```

Otherwise it raises the bar too much for informal discussions, especially for less experienced persons already impressed by public discussions with well known experts.

mbolivar-nordic commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,272 @@
+  mark a PR as draft and explicitly request reviewers to provide early feedback,
+  even with failing CI checks.
+
+- When breaking a PR into multiple commits, each commit must build cleanly. The
+  CI system does not enforce this policy, so it is the PR author's
+  responsibility to verify.
```
No objections to the content, but I do see some overlap between this and https://docs.zephyrproject.org/latest/contribute/guidelines.html#other-commit-expectations

keith-zephyr replied:
Provided a link to this content from the general contribution guidelines.

mbolivar-nordic commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,272 @@
+your PR right away.
+
+The `Zephyr Dev Meeting`_ performs a triage of PRs missing reviewer approval.
+This triage process generally follows this process:
```
nitpick: this reads a little redundant

mbolivar-nordic commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,272 @@
+- After more than 1 week of inactivity, post a message on the `#pr-help`_
+  channel on Discord linking to your PR.
+
+- After 2 weeks of inactivity, add the `dev-review`_ label to your PR. This
```
Note: this requires privileges on GitHub that not all contributors have. Might be worth saying something along the lines of "... or ask someone to do it for you if you can't"

mbolivar-nordic commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,272 @@
+
+.. [#rfc]
+
+   The RFC Proposals doc currently lives under Project Governance. Is there a
+   better location for this doc?
```
Seems like doc/contribute would be better

marc-hb commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,272 @@
+describing the full scope of change and future work.  The RFC proposal provides
+the required context to reviewers, but allows for smaller, incremental, PRs to
+get reviewed and merged into the project. The RFC should also define the minimum
+viable implementation.
```
"It is possible to share and discuss early code supporting RFCs in large, temporary draft PRs that are  too large to be merged as is but intended to be broken down in smaller PRs later."

marc-hb reviewed (commented):
Nothing blocking

mbolivar-nordic reviewed (commented):
Looks good!

I will suggest the following process to get this out of process WG:

- once it's ready, bring to the TSC
- depending on TSC results, make sure to email devel about this to alert the maintainers about this process
- depending on TSC results, ask member representatives to make sure maintainers employed by their companies and internal contributors are aware of the new process
- Once it's done and dusted, courtesy email to users list

fabiobaltieri commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,272 @@
+
+- When breaking a PR into multiple commits, each commit must build cleanly. The
+  CI system does not enforce this policy, so it is the PR author's
+  responsibility to verify.
```
Would not go as far as suggesting that considering how long each run takes. I understand it's just a suggestion but people unfamiliar with the process would end up doing it all the time and burning half a day on it, it sounds intimidating for no good reason (building for one specific case should be good enough).

marc-hb replied:
> Would not go as far as suggesting that considering how long each run takes. [...]  and burning half a day on it, 

This is of course done in _parallel_. 

Clarifying:

"In doubt, this verification can be performed by submitting OTHER, even smaller, temporary, test draft PR(s) without any reviewer and closing them unmerged after they passed the CI tests."

> people unfamiliar with the process would end up doing it all the time [...] it sounds intimidating for no good reason 

"In doubt".

In my fairly extensive experience asking people to submit test PRs, including people unfamiliar with github
- they never look spend time having  extra "fun" with github unless they're asked  or unless it makes sense
- when asked, they all immediately  understand that it can and should be done in a different PR.


More generally speaking, _the practice of tests draft PRs should really be encouraged and developed_. There is a different use case where test draft PRs "save lives" and regularly find (or: could have found!) regressions before merge: when running the CI of one git repo on a code change in a git different repo: 

- https://docs.zephyrproject.org/latest/develop/modules.html#process-for-submitting-changes-to-existing-modules
- https://github.com/zephyrproject-rtos/zephyr/pull/53327#issuecomment-1414193515
- https://github.com/thesofproject/rimage/commit/6e22944e05ea7c6d6c2b61e9632bdbf017644323

fabiobaltieri commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,272 @@
+  push a rebase only, add a comment to the PR indicating which commit is the
+  rebase.
+
+- Request re-reviews once comments are addressed. [#re-review]_
```
I think users with default permissions can't use this, would not bother with this.

keith-zephyr commented:
Commit https://github.com/zephyrproject-rtos/zephyr/commit/20d2c15ff4c233c383b220254a96ad2e03aa936d is a rebase only.  I still need to address outstanding comments.

keith-zephyr commented:
Commit https://github.com/zephyrproject-rtos/zephyr/commit/12ca7f4c05d8cb5126b985179446d704257d7cf0 is another rebase before the resolving the open comments.

keith-zephyr committed:
docs: Move RFCs into a separate document

The RFC proposal documentation better belongs with the other
documentation related to contributing to the Zephyr project.

Signed-off-by: Keith Short <keithshort@google.com>

keith-zephyr commented:
Highlights of https://github.com/zephyrproject-rtos/zephyr/commit/f7aa72f2dd7c0fe7b20cb633792c0d6bc50162a1:
* Changed references of "you" to committer/reviewer as appropriate
* Small PRs -> Smaller PRs
* Provided additional details for when to add new tests (under "Defining Smaller PRs)

alevkoy commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,287 @@
+- PRs must also satisfy all :ref:`merge_criteria` before a member of the release
+  engineering team merges the PR into the zephyr tree.
+
+Maintainers may request committers break up a PR into smaller PRs and may
```
Non-blocking: "Maintainers may request _that_ committers break up a PR into smaller PRs and may request _that_ the commit create an RFC proposal."

alevkoy commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,287 @@
+- :ref:`treewide-changes`.
+- Other large changes that can benefit from the RFC proposal process.
+
+Maintainers have the discretion to request committers create an RFC for PRs that
```
Non-blocking: "that committers create" or "for committers to create"

I think there are a few other instances of missing prepositions in here.

teburd commented:
@keith-zephyr thank you for taking the time to look at and work towards refining the language, this looks pretty good to me now. I will spend some time looking things over one more time before I add a +1, but my concerns I believe have all been addressed.

kartben commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,287 @@
+
+- Reviewed more quickly and reviewed more thoroughly. It's easier for reviewers
+  to set aside a few minutes to review smaller changes several times than it is
+  to allocate large block of time review a large PR.
```
```suggestion
  to allocate large blocks of time to review a large PR.
```

kartben commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,287 @@
+in mind each commit in the PR must still build cleanly and pass all the CI
+tests.
+
+For example, when introducing an extension to an API, Committers can break up
```
```suggestion
For example, when introducing an extension to an API, committers can break up
```

kartben commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,287 @@
+  engineering team merges the PR into the zephyr tree.
+
+Maintainers may request committers break up a PR into smaller PRs and may
+request the committer create an :ref:`RFC proposal <rfcs>`.
```
```suggestion
request they create an :ref:`RFC proposal <rfcs>`.
```

mbolivar-nordic commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,287 @@
+Maintainers may request committers break up a PR into smaller PRs and may
+request the committer create an :ref:`RFC proposal <rfcs>`.
+
+.. _Kernel timer tests: https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/kernel/timer/timer_behavior
+
+.. _test coverage: https://app.codecov.io/gh/zephyrproject-rtos/zephyr/blob/main/kernel/timer.c
+
+.. _`smart battery sensor tests`: https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/sensor/sbs_gauge/src/test_sbs_gauge.c
+
+.. _`smart battery emulator`: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/sbs_gauge/emul_sbs_gauge.c
```
Blocking: please use `:zephyr_file:` to link to files in the tree (grep for examples if unfamiliar). This ensures that release HTML pages point to the corresponding files for that release, unlike direct links, which result in links to files from the future in past release HTML.

kartben commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,287 @@
+  the greater picture.
+
+- Respond to comments using the "Start Review" and "Add Review" green buttons in
+  the "Files changed" view. This allows you to respond to multiple comments and
```
This is the only occurence of "you" in the whole document. At the risk of sounding nitpicky, I'd just remove it from the sentence?

```suggestion
  the "Files changed" view. This allows to respond to multiple comments and
```

mbolivar-nordic commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,287 @@
+
+.. _`smart battery emulator`: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/sbs_gauge/emul_sbs_gauge.c
+
+.. _`fuel gauge API`: https://app.codecov.io/gh/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/fuel_gauge.h
+
+.. _`smart battery sensor driver`: https://app.codecov.io/gh/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/sbs_gauge/sbs_gauge.c
+
```
I'm unsure if these codecov results are usable here for a similar reason.

kartben commented on doc/contribute/proposals_and_rfcs.rst:
```diff
@@ -0,0 +1,55 @@
+
+- A new feature that creates new API surface area, and would require a feature
+  flag if introduced.
+- The modification of an existing stable API
```
```suggestion
- The modification of an existing stable API.
```

kartben commented on doc/contribute/committer_expectations.rst:
```diff
@@ -0,0 +1,287 @@
+  - Fixing a hardware agnostic bug. The test should fail without the bug fixed
+    and pass with the fix applied.
+
+- PRs must update any documentation affected by the functional code changes.
```
```suggestion
- PRs must update any documentation affected by a functional code changes.
```

rerickson1 replied:
`change` not `changes`

kartben reviewed (commented):
Minor spelling/grammar suggestions, feel free to ignore, as I realize I'm late to the party.

Also *very* late, apologies for that, but wouldn't it be worth defining the term "committer"? I see two reasons for that:

- for some open source projects, "committer" has a very different meaning from what is implied here, one that is closer to our definition of maintainer I guess. I would expect that a vast majority of first-time contributors wouldn't consider themselves "committers" and would
- the Guidelines document seems to be referring to the person contributing code subject to the "committer expectation" as a "contributor", so maybe that's the term we should be using? Or maybe "code contributor"? (but maybe not ideal either, and might or might not confuse folks who just contribute e.g. doc fixes).

nashif commented:
> for some open source projects, "committer" has a very different meaning from what is implied here, one that is closer to our definition of maintainer I guess. I would expect that a vast majority of first-time contributors wouldn't consider themselves "committers" and would

that is actually a very good point, goes unnoticed when you are in the weeds. We should stick to the existing terminology we use across the documnetation, which is 'contributor' if I am not mistake.

keith-zephyr commented:
> > for some open source projects, "committer" has a very different meaning from what is implied here, one that is closer to our definition of maintainer I guess. I would expect that a vast majority of first-time contributors wouldn't consider themselves "committers" and would
> 
> that is actually a very good point, goes unnoticed when you are in the weeds. We should stick to the existing terminology we use across the documnetation, which is 'contributor' if I am not mistake.

Sounds good.  I'll make the change from "committer" to "contributor".  This permits me to link to the existing [contributor definition](https://docs.zephyrproject.org/latest/project/project_roles.html#contributor).

Lots of quality feedback is coming in, so I'm going to let the current commit soak a bit and push a new version capturing the outstanding comments by the end of the week.

kartben commented:
> > > for some open source projects, "committer" has a very different meaning from what is implied here, one that is closer to our definition of maintainer I guess. I would expect that a vast majority of first-time contributors wouldn't consider themselves "committers" and would
> > 
> > 
> > that is actually a very good point, goes unnoticed when you are in the weeds. We should stick to the existing terminology we use across the documnetation, which is 'contributor' if I am not mistake.
> 
> Sounds good. I'll make the change from "committer" to "contributor". This permits me to link to the existing [contributor definition](https://docs.zephyrproject.org/latest/project/project_roles.html#contributor).

Great!

> Lots of quality feedback is coming in, so I'm going to let the current commit soak a bit and push a new version capturing the outstanding comments by the end of the week.

Sounds great -- I might actually have another pass then, if I'm not that late in the end :)

marc-hb commented:
> for some open source projects, "committer" has a very different meaning

This meaning is now obsolete but I'm afraid you're right, that old terminology seems to be still sticking in a number of places. Best avoided.

https://en.wikipedia.org/wiki/Committer#Commit_bit
https://www.freebsd.org/internal/proposing-committers

\<gray beard trivia>
In obsolete centralized version control (before the bitkeeper/git/mercurial/... revolution) you cannot commit in private, every commit is public. So commits had to be restricted to a limited number of powerful "committers". I remember very clearly teaching myself `git-svn` for this only purpose: use local commits and branches to manage work in progress privately.

Just for fun, someone thinks commits should be restricted even in private: https://fossil-scm.org/home/doc/trunk/www/rebaseharm.md

keith-zephyr commented:
Note - commit https://github.com/zephyrproject-rtos/zephyr/commit/299ded4fb8bae49c3e353310b61382f19c047d05 only renames the filename committer_expectations to contributor_expectations.  I have another revision to push that will include resolution to the open change requests.

marc-hb commented on doc/contribute/contributor_expectations.rst:
```diff
@@ -0,0 +1,281 @@
+- When using the "Request Changes" option, mark trivial, non-functional,
+  requests as "Non-blocking" in the comment. Reviewers should approve PRs once
+  only non-blocking changes remain. The PR author has discretion as to whether
+  they address all non-blocking comments.
```
```suggestion
  they address all non-blocking comments. Out of respect for the reviewers' time, authors should at least _acknowledge_ every review comment somehow. An emoticon can be enough.
```

marc-hb commented on doc/contribute/contributor_expectations.rst:
```diff
@@ -0,0 +1,281 @@
+  changes in the tree.
+
+- Easier to revert if the PR breaks functionality.
+
```
```suggestion
- This page does not apply to draft PRs which can have any size, any number of commits and any combination of smaller PRs for testing and preview purposes. Draft PRs have no review expectation and PRs created as drafts from the start do not notify anyone by default.
```

marc-hb commented on doc/contribute/contributor_expectations.rst:
```diff
@@ -0,0 +1,281 @@
+  publish the responses in bulk. This reduces the number of emails sent to
+  reviewers.
+
+- Try to minimize rebases in the middle of a review. If a rebase is required,
```
```suggestion
- Try to minimize rebases in the middle of a review as they generate off topic `range-diff` noise. If a rebase is required,
```

marc-hb commented on doc/contribute/contributor_expectations.rst:
```diff
@@ -0,0 +1,280 @@
+  reviewers.
+
+- As GitHub does not implement |git range-diff|_, try to minimize rebases in the
+  middle of a review. If a rebase is required, push this as a separate update
```
> As GitHub does not implement |git range-diff|_, try to minimize rebases in the  middle of a review.

Could an admin please disable the "Update branch" button that shows up in every PR?

It's apparently just one checkbox:
https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/managing-suggestions-to-update-pull-request-branches

stephanosio replied:
> Could an admin please disable the "Update branch" button that shows up in every PR?

Done
